### PR TITLE
semgrep-core: don't filter out explicitly passed targets

### LINKED ISSUE #831: --lang fails to set input file language
**Describe the bug**
The help for `--lang` says “parse pattern and all files in specified language”, but this does not happen (so it’s impossible to lint Python scripts without a `.py` extension):

```console
$ cat .semgrep.yml 
rules:
  - id: test
    languages: [python]
    pattern: bad()
    severity: ERROR
    message: do not call bad()
$ cat bad.py
bad()
$ semgrep --lang=python bad.py
bad.py
ERROR rule:test: do not call bad()
1:bad()
$ mv bad.py bad
$ semgrep --lang=python bad
$
```

**Expected behavior**
```console
$ semgrep --lang=python bad
bad
ERROR rule:test: do not call bad()
1:bad()
```

**Environment**
Official binary (`semgrep-v0.8.0-ubuntu-16.04.tgz`).

### PR DESCRIPTION
**Pull request by brendongo**
Files listed as targets for analysis are analyzed even if they do not
have an expected extension

### CODE REVIEW

File: semgrep-core/bin/Main.ml [hunk 1]
```diff
@@ -415,18 +421,10 @@ let iter_generic_ast_of_files_and_get_matches_and_exn_to_errors f files =
        if !verbose then pr2 (spf "Analyzing %s" file);
        try
          let lang =
-           match Lang.langs_of_filename file with
```
Comment (brendongo) [inline semgrep-core/bin/Main.ml:420]: Remove since this requires files have expected extension

File: semgrep-core/bin/Main.ml [hunk 2]
```diff
@@ -405,7 +405,13 @@ let get_final_files xs =
...
   in
   let files = filter_files files in
-  files
+
+  let explicit_files = xs |> List.filter(fun file ->
+      Sys.file_exists file && not (Sys.is_directory file)
+    )
+  in
+
+  files @ explicit_files
```
Comment (nbrahms) [inline semgrep-core/bin/Main.ml]: How does this avoid double-passing files in each side?

File: semgrep-core/bin/Main.ml [hunk 3]
```diff
@@ -415,18 +421,10 @@ let iter_generic_ast_of_files_and_get_matches_and_exn_to_errors f files =
...
        try
          let lang =
-           match Lang.langs_of_filename file with
-           | [lang] -> lang
-           | x::_xs ->
-               (match Lang.lang_of_string_opt !lang with
-               | Some lang -> lang
-               | None ->
-                 pr2 (spf "no language specified, defaulting to %s for %s"
-                      (Lang.string_of_lang x) file);
```
Comment (nbrahms) [inline semgrep-core/bin/Main.ml:427]: I think you want to move the `langs_of_filename` check in here, and keep the automated detection. So if someone omits `-lang` and passes in `foo.py` we parse as python, but if someone passes `foo.blargh` we error.

File: semgrep-core/bin/Main.ml (hunk 2 above)
Comment (nbrahms) [reply semgrep-core/bin/Main.ml]: FWIW, I'd think you'd make a set out of `files`, and check lack of set membership.

Comment (brendongo) [reply semgrep-core/bin/Main.ml]: Good call this doesn't dedup

File: semgrep-core/bin/Main.ml (hunk 3 above)
Comment (brendongo) [reply semgrep-core/bin/Main.ml:427]: I think it's fine to not work when -lang is not passed?

Comment (nbrahms) [reply semgrep-core/bin/Main.ml:427]: OK, but then I think you'll want to audit all the documentation to make sure we have `-lang` in all our examples.

IMO it's easier to make the smaller change here, but will leave it to you.

Comment (brendongo) [reply semgrep-core/bin/Main.ml:427]: This is to semgrep-core so it's not exposed externally anyway

**Comment by brendongo** [comment]
Tox tests failing since it uses binary built on develop

**Review by DrewDennison — APPROVED** [review]
LGTM

**Pull request merged**